[Andrew Baruch] iP

andrewbaruch opened the pull request:
<h2> Duke </h2>

> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

It is:
- Text Based (CLI)
- Easy to use! kinda...
- ~~Fast~~ *SUPER FAST* :dash:
- Did I mention it is **FREE**

All you need to do is:
1. Install java
2. Download the latest version [here](https://github.com/LeopardMerkava/ip/releases)
3. Launch using your command prompt
4. Enjoy 👍 

Features:
- [X] ToDo
- [X] Events
- [X] Deadlines
- [X] Accepts time date
- [ ] Find (Upcoming!)

And, of course, you can take a look at the code too! Here is the `main`  method (without the comments):
        
```java
public static void main(String[] args) {
        Storage storage = new Storage();
        Ui ui = new Ui();
        Parser parser = new Parser();
        storage.open();
        boolean isLoopBroken = false;
        ui.greet(true);

        while (!isLoopBroken) {
            try {
                Command c = parser.parse(ui.getInput());
                isLoopBroken = c.execute(ui, storage);
            } catch (DukeExceptions e) {
                ui.printException(e);
                isLoopBroken = false;
            }
        }
    }
```

Commit:
Add Gradle support

andrewbaruch committed:
Set Level 1. Added Greeting, Echo, and Exit

andrewbaruch committed:
Change horizontal line to constant

andrewbaruch committed:
Add ability to store input text into a list and display the list

andrewbaruch committed:
Add Task class, Add support for marking task as done

andrewbaruch committed:
Add Deadline, Event, ToDo. Add support for adding deadlines and time.

andrewbaruch committed:
Add Ai Testing, Fix bug on print

andrewbaruch committed:
Add DukeExceptions, Handle most exceptions, Update Expected

andrewbaruch committed:
Add delete command

andrewbaruch committed:
Move constant and variable out of main

Done to make the main code more readable

andrewbaruch committed:
Fix bug, split processing to different functions, overhaul command detect

This step is important as it makes the code much more readable

andrewbaruch committed:
Add FileHandler to make directory and file

andrewbaruch committed:
Add scan for save file function and Add write to save file function

andrewbaruch committed:
Added support to store deadline and time as LocalDateTime

andrewbaruch committed:
Create Command class and it's subclass, UI, Storage, and Parse class.

andrewbaruch committed:
Create duke package to contain all. Create duke.storage, duke.task and
duke.commands for those classes respectively

andrewbaruch committed:
Add Junit test

andrewbaruch committed:
Add Javadocs comment to most of public class and methods

andrewbaruch committed:
Change code to match CodingStandard

andrewbaruch committed:
Add support for find commands

andrewbaruch committed:
Merge branch 'branch-A-CodingStandard'

andrewbaruch committed:
Merge branch 'branch-A-JavaDoc'

lrnshk commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+     */
+
+    public static void main(String[] args) {
+        Storage storage = new Storage();
```
Should all these lines be placed in main? Perhaps using a constructor to initialise, and creating methods to run the various functions in your main method, such as run(), will be more understandable.

lrnshk commented on src/main/java/duke/DukeExceptions.java:
```diff
@@ -0,0 +1,17 @@
+package duke;
+
+<<<<<<< HEAD
```
I think you accidentally committed this file without resolving conflicts! I noticed this in many of your files.

lrnshk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+     * @throws DukeExceptions if the first word of the input is not a recognisable command.
+     */
+    public Command parse(String userInput) throws DukeExceptions {
+        String[] splitUserInput = userInput.split(" ", 2);
```
I like that you named the String array clearly.

lrnshk commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+
+import duke.commands.ByeCommand;
+import duke.commands.Command;
+import duke.commands.DeadlineCommand;
```
I'm not sure you require 1 class for every single command. Perhaps making these comments the functions of classes such as Task, Todo, Deadline, Event etc?

andrewbaruch committed:
Fix Bug and missed conflicts

lrnshk commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * Delete the task in the index.
```
Should "Delete" be "Deletes"?

lrnshk commented on src/main/java/duke/storage/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+            Task task;
+            if (splitString[0].equals("T")) {
+                task = ToDo.create(splitString[2]);
+
```
Is this newline necessary before "} else if (..) {"? Same for the rest of the if-else block

lrnshk commented on src/main/java/duke/storage/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+                    .append(DIVIDER)
+                    .append(currTask.getStatusIcon())
+                    .append(DIVIDER)
+                    .append(currTask.getDescription());
```
Should there be a newline after this line?

lrnshk commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String toString() {
+        return "[" + this.taskType + "]" + "[" + getStatusIcon() + "]" + " " + this.description
```
Would this be more readable if you used String.format for the entire string?

lrnshk reviewed (commented):
Hi, your code looks good! A lot of attention to details, especially with the many scenarios you considered for exceptions. I saw that you pushed a new commit to fix some things, but just note that you missed a lot of conflicts in these files, and I added some code quality improvements.

andrewbaruch committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S1/ip

rebeccacxy commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,75 @@
+     * @param output The message which should be displayed in the output.
+     */
+
+    private static void reply(String output) {
+        System.out.println(HORIZONTAL_LINE);
+        System.out.println(output);
+        System.out.println(HORIZONTAL_LINE);
+    }
```
reply could be renamed printReply?

rebeccacxy commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Prints a string with proper format
+     **/
+    public void print(String s) {
+        reply(s);
+    }
```
Can be renamed better! This method seems to do the same as `reply` so perhaps you could just use one of them?

rebeccacxy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,151 @@
+     *
+     * @param userInput The string which the user typed in.
+     * @return Appropriate command the input represents for the main function to execute.
+     * @throws DukeExceptions if the input is lacking argument.
+     * @throws DukeExceptions if the input has too much argument.
+     * @throws DukeExceptions if the input is not properly formatted.
+     * @throws DukeExceptions if the input is empty.
+     * @throws DukeExceptions if the first word of the input is not a recognisable command.
```
Is this good practice to have so many lines of javadocs for DukeExceptions? 

Personally I feel you could break up the exceptions into separate classes extended from `DukeException`!

rebeccacxy commented on src/main/java/duke/tasks/ToDo.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public class ToDo extends Task{
+
+    private ToDo(String description) {
+        super(description, "T");
+    }
```
Missing javadocs for constructor?

rebeccacxy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    /**
+     * Returns the tsize of the list.
```
Typo —> Should be size

rebeccacxy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    /**
+     * Find task which has the keyword in the list.
```
“Find task” —> Should be Find tasks or Finds a task

rebeccacxy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,144 @@
+     * Find task which has the keyword in the list.
+     *
+     * @param keyword The keyword to look for.
+     * @return String representation of a list of task that has the keyword in the description.
```
List of task —> Should be list of tasks

Sorry, just nitpicking grammar issues :P

rebeccacxy commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class Event extends Task {
+
+    private Event(String description, LocalDateTime time) {
+        super(description, "E", time);
+    }
```
Missing javadocs for constructor?

rebeccacxy commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Deadline extends Task {
+
+    private Deadline(String description, LocalDateTime deadline) {
+        super(description, "D", deadline);
+    }
```
Missing javadocs for constructor?

rebeccacxy commented on data/duke.txt:
```diff
@@ -0,0 +1,6 @@
+T%% %%playgames
+E%% %%play games%%2020-12-21 15:00
+T%% %%read
+T%% %%playgames
+T%% %%play
+T%% %%games
```
Seems like your output is not correctly formatted? Might want to check on this

rebeccacxy reviewed (commented):
Overall not many issues, can see you put in a good amount of effort into writing comprehensive methods. Good job and all the best for the rest of IP :)

andrewbaruch committed:
Add gradle & checkstyle

andrewbaruch committed:
Edit gradle build

andrewbaruch committed:
Add GUI

andrewbaruch committed:
Change code to implement more SLAP

andrewbaruch committed:
Add assertion on MainWindow

dukeImage and userImage may be null if the file is not present.

Putting assertion allows the program to catch this case.

Let's ensure that the program catches when it is null.

andrewbaruch committed:
Parse: Make new methods to handle checks

Parse method is doing too much and is very long.
Some checks are repeated.

Extracting checks to new methods make the code more readable, shorter,
and follows SLAP more closely.

Let's
- take out the repeated checks out to be done by other methods.
- move some of the processing to new methods.

andrewbaruch committed:
Add update feature

andrewbaruch committed:
Merge pull request #1 from LeopardMerkava/A-Assertions

Add assertion on MainWindow

andrewbaruch committed:
Merge pull request #2 from LeopardMerkava/A-CodeQuality

Parse: Make new methods to handle checks

andrewbaruch committed:
Merge pull request #3 from LeopardMerkava/C-Update

Add update feature

andrewbaruch committed:
Fix bug

andrewbaruch committed:
Add new profile picture, fix bugs

andrewbaruch committed:
Change logo, add window name and window logo. Add timer before closing.

andrewbaruch committed:
Change output of find to show the task index

Showing a counting up index causes issue when user try to delete

andrewbaruch committed:
Add bye.png for docs

andrewbaruch committed:
Add Userguide

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Add pictures for userguide

andrewbaruch committed:
Merge branch 'master' of https://github.com/LeopardMerkava/ip

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Set theme jekyll-theme-merlot

andrewbaruch committed:
Set theme jekyll-theme-slate

andrewbaruch committed:
Fix bug

andrewbaruch committed:
Update User Guide

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Set theme jekyll-theme-minimal

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Set theme jekyll-theme-cayman

andrewbaruch committed:
Update README.md

andrewbaruch committed:
Set theme jekyll-theme-slate

andrewbaruch committed:
Add UI Picture
